src/libmeasurement_kit/dns/system_resolver.hpp

```diff
@@ -0,0 +1,174 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include <measurement_kit/dns.hpp>
+
+#include <cassert>
+#include <future>
+
+#include <iostream>
+
+namespace mk {
+namespace dns {
+
+class ResolverContext {
+  public:
+    QueryClass dns_class;
+    QueryType dns_type;
+    std::string name;
+    Callback<Error, Var<Message>> cb;
+    Settings settings;
+    Var<Reactor> reactor;
+    Var<Logger> logger;
+
+    addrinfo hints;
+
+    Var<Message> message{new Message};
+
+    ResolverContext(QueryClass dns_c, QueryType dns_t, std::string n,
+                    Callback<Error, Var<Message>> c, Settings s, Var<Reactor> r,
+                    Var<Logger> l) {
+        dns_class = dns_c;
+        dns_type = dns_t;
+        name = n;
+        cb = c;
+        settings = s;
+        reactor = r;
+        logger = l;
+        memset(&hints, 0, sizeof(hints));
+    }
+};
+
+template <MK_MOCK(getaddrinfo), MK_MOCK(inet_ntop)>
+void resolve_async(ResolverContext *context) {
+    std::unique_ptr<ResolverContext> ctx(context);
+    Callback<Error, Var<Message>> callback = ctx->cb;
+    Var<Message> message = ctx->message;
+    addrinfo *servinfo_p = nullptr;
+
+    int error =
+        getaddrinfo(ctx->name.c_str(), nullptr, &ctx->hints, &servinfo_p);
+    std::unique_ptr<addrinfo, std::function<void(addrinfo *)>> servinfo(
+        servinfo_p, [](addrinfo *servinfo) {
+            if (servinfo != nullptr)
+                freeaddrinfo(servinfo);
+        });
```

Comment:
Also, could you add { } even though there is just one statement in the if branch?